ccm_web/server/src/api/api.enroll.section.users.handler.ts

```diff
@@ -0,0 +1,99 @@
+import CanvasRequestor from '@kth/canvas-api'
+
+import { APIErrorData, isAPIErrorData } from './api.interfaces'
+import { handleAPIError, HttpMethod } from './api.utils'
+
+import baseLogger from '../logger'
+import { CanvasEnrollment } from '../canvas/canvas.interfaces'
+import { SectionUserDto } from './dtos/api.section.users.dto'
+
+const logger = baseLogger.child({ filePath: __filename })
+
+export class EnrollSectionUsersApiHandler {
+  requestor: CanvasRequestor
+  users: SectionUserDto[]
+  sectionId: number
+
+  constructor (requestor: CanvasRequestor, users: SectionUserDto[], sectionId: number) {
+    this.requestor = requestor
+    this.users = users
+    this.sectionId = sectionId
+  }
+
+  makeResponse (enrollmentResult: Array<APIErrorData | CanvasEnrollment>): CanvasEnrollment[] | APIErrorData {
+    const failures = []
+    const statusCodes: Set<number> = new Set()
+    const successes = []
+
+    for (const enrollment of enrollmentResult) {
+      if (isAPIErrorData(enrollment)) {
+        const {
+          statusCode,
+          errors
+        } = enrollment
+        failures.push(...errors)
+        statusCodes.add(statusCode)
+      } else {
+        successes.push(enrollment)
+      }
+    }
+
+    if (successes.length === this.users.length) {
+      return successes
+    } else {
+      return {
+        statusCode: statusCodes.size > 1 ? 502 : [...statusCodes][0],
+        errors: failures
+      }
+    }
+  }
```

Comment:
I think we could make this a utility function in `api.utils.ts`, since it's basically doing the same thing as what is being done in `api.create.section.handler.ts `. To do that, we probably need to use a [generic](https://www.typescriptlang.org/docs/handbook/2/generics.html). I think the method signature would look something like this:

```
makeResponse<T extends CanvasEntity> (inputCount: number, result: Array<APIErrorData | T>): T[] | APIErrorData { ...}
```

where `CanvasEntity` is defined in `canvas.interfaces.ts` as

```
type CanvasEntity = CanvasEnrollment | CanvasSection | CanvasCourse ...
```

It might need some additional finessing, but seems like a common enough operation that we should extract it, provided we can get the typing to work (which we should be able to). 

Side note: I am thinking of making `api.utils` into another service, so we can inject other services into it if necessary (like `CanvasService`, if we need to modify tokens based on failures for example). In that case, this new function would just be a method.

Reply:
Yes, I got some warnings from IntelliJ IDEA about duplicated code.  It'd make sense to generalize this and move it into the utils.

Reply:
It might make more sense to make this a base class for something that implements a partial success response.

If that's possible with generics, anyways.

-

Reply:
I'm working on implementing the generic.  I think we won't need an argument like `inputCount: number`.  I'll change the logic to check for 0 errors rather than a full list of successes.  I think that's equivalent for our purposes.

Reply:
@ssciolla & @chrisrrowland: I'm generalizing `makeResponse` in `api.utils` for use by class `EnrollSectionUsersApiHandler`.

Should I also update the section class, `CreateSectionApiHandler`, to use it, too?  Or should that be a different issue?

I'm leaning towards opening a new issue for it because:

* It doesn't fit logically in this PR for bulk enrollment.
* If it were in this PR, it would mean that the bulk enrollment **_AND_** section creation API entries should be tested before this PR could be merged.  That seems like a lot of testing.

Reply:
Seems like it makes sense as a separate issue to me.

Reply:
I guess I'm okay either way here, but my inclination would be to just use it in both places now, as part of testing this new function is to make sure it will work in both places. I don't think there's a very big risk in doing so, nor do I think it would be that difficult to test.